sdks/python/apache_beam/runners/render.py

```diff
@@ -0,0 +1,529 @@
+        yield from self.transform_to_dot(
+            subtransform, pcoll_leaf_consumers, edges_out)
+      yield '}'
+
+  def transform_node(self, transform_id):
+    return f'"{transform_id}" [{self.transform_attributes(transform_id)}]'
+
+  def transform_attributes(self, transform_id):
+    transform = self.pipeline.components.transforms[transform_id]
+    local_name = transform.unique_name.split('/')[-1]
+    if transform_id in self.latest_positions:
+      pos_str = f'pos="{self.latest_positions[transform_id]}"'
+    else:
+      pos_str = ''
+    return (
+        f'label="{local_name}" {self.style(transform_id)} '
+        f'URL="javascript:click(\'{transform_id}\')" {pos_str}')
+
+  def pcoll_leaf_consumers_iter(self, transform_id):
+    transform = self.pipeline.components.transforms[transform_id]
+    transform_inputs = set(transform.inputs.values())
+    side_inputs = set(translations.side_inputs(transform).values())
+    if self.is_leaf(transform_id):
+      for pcoll in transform.inputs.values():
+        yield pcoll, (transform_id, pcoll in side_inputs)
+    for subtransform in transform.subtransforms:
+      for pcoll, (consumer,
+                  annotation) in self.pcoll_leaf_consumers_iter(subtransform):
+        if self.is_leaf(transform_id):
+          if pcoll not in transform_inputs:
+            yield pcoll, (transform_id, annotation)
+        else:
+          yield pcoll, (consumer, annotation)
+
+  def pcoll_leaf_consumers(self):
+    result = collections.defaultdict(list)
+    for transform_id in self.roots:
+      for pcoll, consumer_info in self.pcoll_leaf_consumers_iter(transform_id):
+        result[pcoll].append(consumer_info)
+    return result
+
+  def is_leaf(self, transform_id):
+    return (
+        transform_id in self.leaf_composites or
+        not self.pipeline.components.transforms[transform_id].subtransforms)
+
+  def info(self):
+    if len(self.highlighted) != 1:
+      return ''
+    transform_id = self.highlighted[0]
+    return f'<pre>{self.pipeline.components.transforms[transform_id]}</pre>'
+
+  def layout_dot(self):
+    with open('out.dot', 'w') as fout:
+      fout.write(self.to_dot())
+    layout = subprocess.run(['dot', '-Tdot'],
+                            input=self.to_dot().encode('utf-8'),
+                            capture_output=True,
+                            check=True).stdout
+
+    # Try to capture the positions for layout consistency.
+    json_out = json.loads(
+        subprocess.run(['dot', '-n2', '-Kneato', '-Tjson'],
+                       input=layout,
+                       capture_output=True,
+                       check=True).stdout)
+    for box in json_out['objects']:
+      name = box.get('name', None)
+      if box.get('name', None) in self.pipeline.components.transforms:
+        if 'pos' in box:
+          self.latest_positions[name] = box['pos']
+        elif 'bb' in box:
+          x0, y0, x1, y1 = [float(r) for r in box['bb'].split(',')]
```

Comment:
lint fails here due to unused variables.

Reply:
Fixed.

Reply:
the expression is incorrectly quoted. did you mean to supply an f-string?